### PR TITLE
Common temporary location manager for profiling product

### PR DESCRIPTION
**Pull request by jbachorik**
# What Does This Do
It extracts the functionality related to temporary location cleanup (previously only for JFR) to allow for shared experience for JFR recording, ddprof recording and temporary library artifacts.

# Motivation
Unification of the temporary location handling. Providing cleanup of all orphaned temporary artifacts generated by the profiler.

# Additional Notes
## How does this work
The temp location manager will create one common root for the temporary data and that location will have the name of `pid_<pid>` where `<pid>` is the Java process PID. Then, the users can request temporary subfolders with a specific path and these will be created under the common, process-specific, location. All the temporary subfolders are automatically registered for cleanup on JVM exit.
In order to handle hard-kills and JVM crashes, each time a profiled process is started it will scan the base temporary location for any `pid_<pid>` subfolders that don't correspond to any of the currently running JVM. There is an additional requirement for a file/directory to be cleaned up, though. It must have been last modified at least 2*<upload period> seconds ago. This helps to exclude JVMs being currently in shutdown and therefore not showing up in `jps` output or having been created between the cleaner obtained the `jps` output and started cleaning the files.

I had to modify `PidHelper` to drain the pipes asynchronously - apparently, in smoke tests and system tests the number of java processes running on a single host was large enough to cause clogging and effectively deadlocking the test process with the 'jps' subprocess.

_Compatibility issues_
- The configuration flag `profiling.jfr.repository.cleanup` is removed and ignored; cleanup will happen always
- The configuration flag `profiling.jfr.repository.base` is deprecated and ignored; use `profiling.tempdir` instead

# Contributor Checklist

- [x] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [x] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [x] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [x] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [ ] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [PROF-10925]

[PROF-10925]: https://datadoghq.atlassian.net/browse/PROF-10925?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/TempLocationManager.java [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+    tempDir = baseTempDir.resolve("pid_" + pid);
+    cleanupTask = CompletableFuture.runAsync(() -> cleanup(false));
+
+    Runtime.getRuntime()
+        .addShutdownHook(
+            new Thread(
+                () -> {
+                  try {
+                    cleanupTask.join();
```
Comment (richardstartin) [inline dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/TempLocationManager.java:224]: Can this block JVM shutdown? Can we put a timeout here?

**Review by richardstartin — COMMENTED** [review]
Looks good, I just think we need to be sure we don't make the JVM hang on shutdown

File: dd-java-agent/agent-profiling/profiling-controller/src/test/java/com/datadog/profiling/controller/TempLocationManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    tempLocationManager.cleanup(false);
+    // fake temp location should be deleted
+    // real temp location should be kept
+    assertFalse(Files.exists(fakeTempDir));
+    assertTrue(Files.exists(tempDir));
+  }
+
+  @ParameterizedTest
+  @ValueSource(strings = {"preVisitDirectory", "visitFile", "postVisitDirectory"})
+  void testConcurrentCleanup(String section) throws Exception {
```
Comment (manuel-alvarez-alvarez) [inline dd-java-agent/agent-profiling/profiling-controller/src/test/java/com/datadog/profiling/controller/TempLocationManagerTest.java:115]: :heart: nice test!

File: dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/TempLocationManager.java [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    }
+  }
+
+  private final Path baseTempDir;
+  private final Path tempDir;
+  private final long cutoffSeconds;
+
+  private final CompletableFuture<Void> cleanupTask;
+
+  private final CleanupHook cleanupTestHook;
```
Comment (manuel-alvarez-alvarez) [inline dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/TempLocationManager.java:144]: It's a pity that this leaked in the code, we could add a protected method that constructs the `CleanupVisitor` and then override it in the test with the required synchronization (I'm OK with this anyway)

File: dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/TempLocationManager.java (hunk 1 above)
Comment (jbachorik) [reply dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/TempLocationManager.java]: Yes, you are right. If we are unlucky and there is an issue with filesystem or the temp location is on some remote network location (seen stranger things ...) we might block the shutdown.
I am adding support for timeout in the cleanup task itself as well as here in the shutdown routine.

**Review by richardstartin — APPROVED** [review]

**Pull request merged**